src/test/java/org/elasticsearch/indices/recovery/RecoveryStateTest.java

```diff
@@ -481,4 +484,46 @@ Start createObj() {
             assertThat(lastRead.time(), lessThanOrEqualTo(start.time()));
         }
     }
+
+    @Test
+    public void testConcurrentAccessToIndexFileDetails() throws InterruptedException {
```

Comment:
It was not removing entries from the map or adding any while the streamer wrote. The exception came from the call to AbstractCollection.toArray in writeTo

http://build-us-00.elastic.co/job/es_core_master_strong/3228/testReport/junit/org.elasticsearch.recovery/RelocationTests/testMoveShardsWhileRelocation/